### PR TITLE
i18n: refactor TranslationRef and rework TranslationMessages and TranslationResource

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

This replaced #19858, since there was an infinite render loop in that one which is fixed by this change, and there are also a few additional changes to `TranslationRef`. Reasoning for that part of the refactor remains the same.

For the rest this is what I think is a pretty good initial structure for how to provide messages overrides and translations. Overall the structure is as follows:

- A `TranslationRef` provides default messages, and may also optionally provide lazy loaded translations for other languages. The additional translations are type safe as they must match the default message keys.
- For overriding default messages you can pass a `TranslationMessages` to the app, created via `createTranslationMessages`. This is synchronous, it does not use lazy loading.
- For providing translations you pass `TranslationResource` to the app, created via `createTranslationResource`. All translations must be lazy loaded. They can be defined in `.json` files or TypeScript modules. When defined as TypeScript modules the `createTranslationMessages` helper with `full: true` can be used to make sure all message keys are covered.

Or the above in code...

```ts
// plugin
const testRef = createTranslationRef({
  id: 'test',
  messages: { foo: 'foo' },
  translations: {
    de: () => import('./translations/de.json'),
    fr: () => import('./translations/fr'),
  }
})

// app
createApp({
  translations: {
    resources: [
      createTranslationMessages({ ref: testRef, messages: { foo: 'Foo' } }), // Override
      createTranslationResources({
        ref: testRef,
        translations: {
          sv: () => import('./translations/test/sv'),
        }
      })
    ]
  }
})

// app/translations/test/sv
export default createTranslationMessages({
  ref: testRef,
  full: true,
  messages: {
    foo: 'Föö',
  }
})
```

Worth noting here is that when creating a new translation ref it can not be references in any of its options, including in any of the translation files. That means that those files can't use the `createTranslationMessages()` helper, but you'll still get a type error if keys are missing.

You can also set messages values to `null` to fall back to the default message in translations, figuring that's useful for being able to add new messages without initial translations.

CC @mario-mui 

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/core-app-api/alpha-api-report.md [hunk 1]
```diff
@@ -5,32 +5,18 @@
...
-  addLazyResources<Messages extends Record<string, string>>(
-    translationRef: TranslationRef<Messages>,
-    initResources?: Record<
-      string,
-      () => Promise<{
-        messages: TranslationMessages<TranslationRef>;
-      }>
-    >,
-  ): void;
+  addLazyResources(resource: TranslationResource): void;
```
Comment (freben) [inline packages/core-app-api/alpha-api-report.md:15]:
```suggestion
  addLazyResource(resource: TranslationResource): void;
```

should this be singular?

File: packages/core-app-api/alpha-api-report.md [hunk 2]
```diff
@@ -5,32 +5,18 @@
...
-      string,
-      TranslationMessages<TranslationRef<Messages>>
-    >,
-  ): void;
+  addMessages(messages: TranslationMessages): void;
   // (undocumented)
-  addResourcesByRef<Messages extends Record<string, string>>(
-    translationRef: TranslationRef<Messages>,
-  ): void;
+  addResource(translationRef: TranslationRef): void;
```
Comment (freben) [inline packages/core-app-api/alpha-api-report.md:19]: tempting to just make this take `TranslationRef | TranslationResource` ... of hey, a single method taking `TranslationRef | TranslationResource | TranslationMessages`. From a consumer standpoint it doesn't provide much value to have these as three different methods, and one being about lazy stuff isn't really valuable to discern for, from the user's POV

File: packages/core-plugin-api/src/translation/TranslationRef.ts [hunk 3]
```diff
@@ -14,48 +14,133 @@
...
+  TMessages extends { [key in string]: string } = { [key in string]: string },
+> extends TranslationRef<TId, TMessages> {
+  version: 'v1';
+
+  getDefaultMessages(): AnyMessages;
+
+  getDefaultResource(): TranslationResource | undefined;
+}
+
+export type NoInfer<T> = T extends infer S ? S : never;
```
Comment (freben) [inline packages/core-plugin-api/src/translation/TranslationRef.ts:49]: remove?

File: packages/core-app-api/alpha-api-report.md (hunk 1 above)
Comment (Rugvip) [reply packages/core-app-api/alpha-api-report.md:15]: Yeah but I opted for minimizing changes, looking to remove it completely in a followup.

File: packages/core-app-api/alpha-api-report.md (hunk 2 above)
Comment (Rugvip) [reply packages/core-app-api/alpha-api-report.md:19]: Plan is to make this a `getX(ref: TranslationRef)` and not have the resources or messages be referenced directly here at all. Where X is something that lets you generate translations :p

**Comment by mario-mui** [comment]
Honestly I think this idea is nice, means that developer maybe have two purposes to use i18n

1. first is overwrite default messages you known there were some PRS to define some attribute for plugin to customize some titles or descriptions 
2. really need to i18n backstage for their natives

maybe first purpose only need `createTranslationMessages` with few messages. Rear purpose need use `createTranslationResources ` with huge messages(lazy load is necessary)

**Review by freben — APPROVED** [review]
Still no need for NoInfer, but approv'd

**Pull request merged**